cpp/src/arrow/compute/exec/spilling_join.cc

```diff
@@ -0,0 +1,351 @@
+
+            for(int64_t i = 0; i < num_rows; i++)
+            {
+                uint64_t hash = hashes[i];
+                size_t partition = hash & (SpillingAccumulationQueue::kNumPartitions - 1);
+                bool found = partitions[partition] ? partitions[partition]->Find(hash) : true;
+                bit_util::SetBitTo(bv, i, found);
+            }
+        }
+
+        Status SpillingHashJoin::Init(
+            QueryContext *ctx,
+            JoinType join_type,
+            size_t num_threads,
+            SchemaProjectionMaps<HashJoinProjection> *proj_map_left,
+            SchemaProjectionMaps<HashJoinProjection> *proj_map_right,
+            std::vector<JoinKeyCmp> *key_cmp,
+            Expression *filter,
+            PartitionedBloomFilter *bloom_filter,
+            CallbackRecord callback_record,
+            bool is_swiss)
+        {
+            ctx_ = ctx;
+            num_threads_ = num_threads;
+            callbacks_ = std::move(callback_record);
+            bloom_filter_ = bloom_filter;
+            is_swiss_ = is_swiss;
+
+            HashJoinImpl::CallbackRecord join_callbacks;
+            join_callbacks.register_task_group = callbacks_.register_task_group;
+            join_callbacks.start_task_group = callbacks_.start_task_group;
+            join_callbacks.output_batch = callbacks_.output_batch;
+            join_callbacks.finished = [this](int64_t num_total_batches)
+            {
+                return this->OnCollocatedJoinFinished(num_total_batches);
+            };
+
+            builder_ = BloomFilterBuilder::Make(
+                num_threads_ == 1
+                ? BloomFilterBuildStrategy::SINGLE_THREADED
+                : BloomFilterBuildStrategy::PARALLEL);
+            RETURN_NOT_OK(build_accumulator_.Init(ctx));
+            RETURN_NOT_OK(probe_accumulator_.Init(ctx));
+
+            for(size_t i = 0; i < SpillingAccumulationQueue::kNumPartitions; i++)
+            {
+                ARROW_ASSIGN_OR_RAISE(impls_[i], is_swiss_ ? HashJoinImpl::MakeSwiss() : HashJoinImpl::MakeBasic());
+                RETURN_NOT_OK(impls_[i]->Init(ctx_,
+                                              join_type,
+                                              num_threads,
+                                              proj_map_left,
+                                              proj_map_right,
+                                              key_cmp,
+                                              filter,
+                                              join_callbacks));
+
+                task_group_bloom_[i] = callbacks_.register_task_group(
+                    [this](size_t thread_index, int64_t task_id)
+                    {
+                        return PushBloomFilterBatch(thread_index, task_id);
+                    },
+                    [this](size_t thread_index)
+                    {
+                        partition_idx_++;
```

Comment:
Would be nice to move the code to OnNextBloomFilterFinished, then we can just look at On... functions to see the full chain of events and steps in hash join.

Reply:
Moved it